Depthwise3x3s1relu6 optimize

xingjing1 opened the pull request:
优化Depthwise3x3s1relu6，性能在v7和v8均有提升。
![image](https://user-images.githubusercontent.com/33440396/120787652-26999b80-c562-11eb-91fa-e521278976ed.png)

xingjing1 committed:
add conv3x3s1_depthwise_relu6

xingjing1 committed:
conv_impl.cc change

xingjing1 committed:
DWconvrelu6 not trans weight

xingjing1 committed:
fix compile bug asm oprend > 30

chenjiaoAngel commented on lite/backends/arm/math/conv3x3s1_depthwise_int8.cc:
```diff
@@ -29,6 +29,505 @@ namespace math {
 
+#ifdef __aarch64__
+#define INIT_INT8_S1                     \
+                                         \
```
这个空行delete

xingjing1 replied:
好的，已修改

chenjiaoAngel commented on lite/backends/arm/math/conv3x3s1_depthwise_int8.cc:
```diff
@@ -4204,236 +3861,2401 @@ void conv_depthwise_3x3s1p1_bias_relu_int8_int8(int8_t* dout,
+#ifdef __aarch64__
+                asm volatile(
+                    "PRFM PLDL1KEEP, [%[din_ptr0]] \n"
+                    "PRFM PLDL1KEEP, [%[din_ptr1]] \n"
```
这个地方怎么没有调用前面的宏定义呢？

xingjing1 replied:
已修改

chenjiaoAngel commented on lite/backends/arm/math/conv3x3s1_depthwise_int8.cc:
```diff
@@ -4204,236 +3861,2401 @@ void conv_depthwise_3x3s1p1_bias_relu_int8_int8(int8_t* dout,
+                    "PRFM PLDL1KEEP, [%[din_ptr2]] \n"
+                    "PRFM PLDL1KEEP, [%[din_ptr3]] \n"
+                // middle
+                    "ld1    {v0.8b}, [%[din_ptr0]], #8               \n"   /* load a(0,0)-a(0,7) to q0*/
```
同上，这个可以调用Middle部分内容

xingjing1 replied:
已修改

chenjiaoAngel commented on lite/backends/arm/math/conv3x3s1_depthwise_int8.cc:
```diff
@@ -6025,6 +7779,10 @@ void conv_depthwise_3x3s1p0_bias_relu_int8_int8(int8_t* dout,
                     "vmla.f32 q15, q9, q7                            \n"
                     "vmax.f32 q14, q14, q1              \n"
                     "vmax.f32 q15, q15, q1              \n"
+                    "vld1.32 {d2-d3}, [%[alpha]]            \n"
```
V7 汇编也存在重复地方，建议当前PR或下个PR也进行修复下

xingjing1 replied:
下个pr修复

chenjiaoAngel commented on lite/backends/arm/math/conv3x3s1_depthwise_int8.cc:
```diff
@@ -29,6 +29,505 @@ namespace math {
+  "st1    {v6.8b}, [%[ptr_out0]], #8      \n" \
+  "st1    {v8.8b}, [%[ptr_out1]], #8      \n"
+#endif
+
```
这块宏定义的命名建议修改下，改成这样：INIT_S1、LEFT_COMPUTE_S1、LEFT_RESULT_S1、LEFT_RESULT_S1_RELU、LEFT_RESULT_S1_RELU6 这种

xingjing1 replied:
已修改

chenjiaoAngel reviewed (commented):
3x3s1p0 的实现汇编部分可以套用3x3s1p1的MID_COMPUTE_MID_COMPUTE_S1部分，这样减少代码量；第二个V7 汇编部分可以在这个PR或下个PR类似V8方式进行整理；这块内容的整理和宏定义命名可参考：[3x3s1p01_fp32 实现](https://github.com/PaddlePaddle/Paddle-Lite/blob/develop/lite/backends/arm/math/conv3x3s1p01_depthwise_fp32.cc)或[3x3s1_fp16实现](https://github.com/PaddlePaddle/Paddle-Lite/blob/develop/lite/backends/arm/math/fp16/conv3x3s1_depthwise_fp16.cc)

chenjiaoAngel commented on lite/backends/arm/math/conv3x3s1_depthwise_int8.cc:
```diff
@@ -29,6 +29,505 @@ namespace math {
+  "fmla   v19.4s, v13.4s, %[v_scale].4s          \n"
+
+#define MIDDLE_FLOAT \
+  STORE_MIDDLE_FLOAT \
```
STORE_MIDDLE_FLOAT 这个应该是STORE_LEFT_FLOAT吧

xingjing1 replied:
已修改

chenjiaoAngel commented on lite/backends/arm/math/conv3x3s1_depthwise_int8.cc:
```diff
@@ -29,6 +29,505 @@ namespace math {
+  MIDDLE
+
+#define MIDDLE_INT8 \
+  STORE_MIDDLE_INT8 \
```
这个同上

xingjing1 replied:
已修改

chenjiaoAngel commented on lite/backends/arm/math/conv3x3s1_depthwise_int8.cc:
```diff
@@ -29,6 +29,505 @@ namespace math {
+  "fmla   v9.4s, v13.4s, %[v_scale].4s     \n"
+
+#define RIGHT_FLOAT \
+  STORE_RIGHT_FLOAT \
```
这个同上，他应该是STORE_MIDDLE_FLOAT

xingjing1 replied:
已修改

xingjing1 committed:
clean code

xingjing1 committed:
add test

xingjing1 committed:
fix bugs

clean code

add test

change conv_depthwise.cc

simplify code with define

clean conv_depthwise.cc

[conv_int8_compute_test.cc] clean code

clean code

clean code

clean code

chenjiaoAngel reviewed (approved):
LGTM
